File: EBookShopDI/app/src/main/java/com/androidtutz/anushka/ebookshop/viewmodel/factory/MainViewModelFactory.java
package com.androidtutz.anushka.ebookshop.viewmodel.factory;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.androidtutz.anushka.ebookshop.model.EBookShopRepository;
import com.androidtutz.anushka.ebookshop.viewmodel.MainActivityViewModel;

import javax.inject.Inject;
import javax.inject.Singleton;

@SuppressWarnings("ALL")
@Singleton
public class MainViewModelFactory implements ViewModelProvider.Factory {

    private EBookShopRepository repository;

    @Inject
    public MainViewModelFactory(EBookShopRepository repository) {
        this.repository = repository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if(modelClass.isAssignableFrom(MainActivityViewModel.class)) {
            return (T) new MainActivityViewModel(repository);
        }
        throw new IllegalArgumentException("Trolhei!!!");
    }
}

File: ViewModelDemo1UpgradeLesson/app/src/main/java/com/anushka/viewmodeldemo1/viewmodel/factory/MainViewModelFactory.kt
package com.anushka.viewmodeldemo1.viewmodel.factory

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.anushka.viewmodeldemo1.viewmodel.MainActivityViewModel
import java.lang.IllegalArgumentException

@Suppress("UNCHECKED_CAST")
class MainViewModelFactory(private val initial: Int): ViewModelProvider.Factory {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if(modelClass == MainActivityViewModel::class.java) {
            return MainActivityViewModel(initial) as T
        }

        throw IllegalArgumentException("God dammit!")
    }
}